0-common/igluctl/src/main/scala/com.snowplowanalytics.iglu/ctl/commands/Pull.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2012-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+*/
+
+package com.snowplowanalytics.iglu.ctl.commands
+
+import java.nio.file.{Path, Paths}
+import java.util.UUID
+
+import cats.data.{EitherT, NonEmptyList}
+import cats.effect.IO
+import cats.implicits._
+
+import fs2.Stream
+
+import scalaj.http.{Http, HttpRequest, HttpResponse}
+
+import org.json4s.{DefaultFormats, JValue}
+import org.json4s.jackson.JsonMethods.{parse => jacksonParse}
+
+import com.snowplowanalytics.iglu.ctl.commands.Push.{HttpUrl, buildCreateKeysRequest, deleteKey, getApiKeys} // TODO: Spin off into separate package object so we don't have to import from Push
+import com.snowplowanalytics.iglu.ctl.{Common, Result => IgluctlResult}
+import com.snowplowanalytics.iglu.ctl.File._
+import com.snowplowanalytics.iglu.core.SelfDescribingSchema.{parse => igluParse}
+import com.snowplowanalytics.iglu.core.json4s.implicits._
+
+/**
+ * Companion object, containing functions not closed on `masterApiKey`, `registryRoot`, etc
+ */
+object Pull {
+
+  // json4s serialization
+  private implicit val formats = DefaultFormats
+
+  /** Primary function, performing IO reading, processing and printing results */
+  def process(registryRoot: HttpUrl,
+              outputDir: Path,
+              masterApiKey: UUID): IgluctlResult = {
+    val createKeysRequest = buildCreateKeysRequest(registryRoot, masterApiKey)
+    val acquireKeys = Stream.bracket(getApiKeys(createKeysRequest)) { keys =>
+      val deleteRead = deleteKey(registryRoot, masterApiKey, keys.read, "read")
+      val deleteWrite = deleteKey(registryRoot, masterApiKey, keys.write, "write")
+      EitherT.liftF(deleteWrite *> deleteRead)
+    }
+
+    val schemas = (for {
+      keys <- acquireKeys
+    } yield getSchemas(buildPullRequest(registryRoot, keys.read)))
+      .compile.toList.getOrElse(throw new RuntimeException("Could not get schema."))
+      .map(_.map(_.value)).map(_.traverse(effect => EitherT(effect)).value).flatten
+
+    val result = schemas.map(_.getOrElse(throw new RuntimeException("Could not write schema.")) //Should this be throwing an exception? If 1 out of 10 schemas does not validate, should we not write the other 9 and silently drop the invalid one?
```

Comment:
But in fact whole program can be simplified even further:

```scala
    val schemasStream = for {
      keys <- acquireKeys
      response = getSchemas(buildPullRequest(registryRoot, keys.read))
    } yield response.leftMap(message => Common.Error.ServiceError(message): Common.Error)

    val result = for {
      response <- schemasStream.compile.toList.flatMap(_.sequence)
      output <- response.traverse(r => writeSchemas(r, outputDir))  // in fact this is always single-element list
    } yield output.flatten

    result.leftMap(error => NonEmptyList.of(error))
```

The only one useful function here is `writeSchema`, everything is just getting around types.

(And ideally nothing should be throwing exceptions)